BIP125: Enable replace-by-fee in mempool

pinheadmz opened the pull request:
This is an initial implementation of mempool replace-by-fee as described in [BIP125:](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki)

> 1. The original transactions signal replaceability explicitly or through inheritance as described in the above Summary section. (nSequence number less than (0xffffffff - 1))
> 2. The replacement transaction may only include an unconfirmed input if that input was included in one of the original transactions. (An unconfirmed input spends an output from a currently-unconfirmed transaction.)
> 3. The replacement transaction pays an absolute fee of at least the sum paid by the original transactions.
> 4. The replacement transaction must also pay for its own bandwidth at or above the rate set by the node's minimum relay fee setting. For example, if the minimum relay fee is 1 satoshi/byte and the replacement transaction is 500 bytes total, then the replacement must pay a fee at least 500 satoshis higher than the sum of the originals.
> 5. The number of original transactions to be replaced and their descendant transactions which will be evicted from the mempool must not exceed a total of 100 transactions.

My implementation is based on Bitcoin Core PR https://github.com/bitcoin/bitcoin/pull/6871 with one subtle difference: Bitcoin Core has an extra rule that rejects replacement transactions that _spend an output they would be replacing._ That also applies to outputs of descendant transactions that would be evicted from the mempool if their parent was replaced. Obviously this is an invalid transaction because its acceptance in to the mempool would cause its own input to stop existing. However, rule # 2 above already catches this edge case. The only reason it is explicit in Bitcoin Core is because rule # 2 _could be_ bypassed by policy configuration whereas the extra rule applies to a subset of absolutely invalid transactions. It's a bit of a tricky check because ancestors and conflicts all need to be cross-compared, and I don't think we need it.

Options like this are [apparently available in Bitcoin Knots.](https://github.com/bitcoin/bitcoin/pull/7219)

We should also discuss whether to eliminate the `replace-by-fee` option altogether. It is not a config option in Bitcoin Core. Maybe this can be discussed after the bcoin wallet is ready for RBF.

tynes commented on lib/mempool/mempool.js:
```diff
@@ -844,7 +844,20 @@ class Mempool extends EventEmitter {
+    // Remove conflicting TXs and their descendants
+    if (conflicts.length) {
+      assert(this.options.replaceByFee);
+
```
Its not immediately obvious that `this.verify` will return an empty list if `this.options.replaceByFee` is turned off. Now you need to know implementation specific details of `this.verify` in this context to be sure that the `assert` will never trigger. The same code could be expressed in what feels like a safer way:

```js
if (this.options.replaceByFee) {
    for (const conflict of conflicts) {
      this.logger.debug(
          'Replacing tx %h with %h',
          conflict.tx.hash(),
          tx.hash());
      this.evictEntry(conflict);
    }
}
```

pinheadmz replied:
Gotcha, this is good feedback thanks. I think we should consider actually just removing `replaceByFee` as an option altogether, like Bitcoin Core.

Also, the options-checks aren't entirely necessary except for the first one here (not affected by this PR):

https://github.com/bcoin-org/bcoin/blob/cec3c3e788834d3e9fc3c0b21bc3e340d573dcfd/lib/mempool/mempool.js#L784-L790

After this point, since all opt-in-RBF TXs are rejected, there will never be any `conflicts` because they would be rejected as double spends. So, the `assert` statement here could be removed. All the other `options.replaceByFee` checks are just optimizations, skipping tests that won't apply if there aren't any RBF TXs in the mempool anyway.

tynes replied:
I'd vote in favor of leaving it, optimizing for more optionality in the ecosystem. Unless we decide that it would cause a great amount of implementation complexity, which I don't think will.

pinheadmz replied:
Ok I believe this is more clear now: https://github.com/bcoin-org/bcoin/pull/811/commits/3d5ff9b842497bbf9699d83715f5ccc121fe8ea9

tynes commented:
I'd personally like to run `bcoin` with RBF enabled, so taking another look at this. One thought is that we should also be sure that the wallet does not index "zero conf" txs as confirmed if they are RBF.  I think that the wallet changes should be merged in at the same time to the mempool changes.

The relevant codepath in the wallet:
https://github.com/bcoin-org/bcoin/blob/master/lib/wallet/txdb.js#L417

tynes commented on lib/mempool/mempool.js:
```diff
@@ -1659,13 +1741,36 @@ class Mempool extends EventEmitter {
+      const {hash, index} = prevout;
+      const conflict = this.getSpent(hash, index);
+
+      if (conflict && conflicts.indexOf(conflict) === -1)
```
It would be better to use a `Set` instead of an array for `conflicts` in this case, `indexOf` will need to iterate through the whole array each time looking for the `conflict`. The `Set` can be converted to an array when its being returned.

pinheadmz replied:
This is good feedback as well - but the function `getConflicts()` does actually need to return an array of `MempoolEntries`. The caller, `verify()` checks the entries' fee rates and in-mempool descendants, so a `BufferSet` of hashes would just mean getting those entries later anyway, possibly in multiple places.

pinheadmz replied:
I guess I could use a `BufferSet` just internally for this function, like you said to reduce search time, but still keep an array of actual `MempoolEntries` to return?

tynes replied:
You can use `Array.from(set)` to convert a set to an array

pinheadmz replied:
updated here: https://github.com/bcoin-org/bcoin/pull/811/commits/2a2c3de12fa7e1b3e46e98d7440535ac1576ac7f

tynes commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+    if (this.options.replaceByFee)
+      conflicts = this.getConflicts(tx);
+
+    if (conflicts.length) {
```
I feel like this `if` statement is boilerplate code in the sense where it just adds additional line that a reviewer needs to read and moves everything in its block two spaces over when the for loop below (`for (const conflict of conflicts)`) would just never run if `conflicts.length` is `0`

pinheadmz replied:
Hm thanks. The `if` does cut off a lot of unnecessary checks however, not just the `for` loop. Especially the `replacement-adds-unconfirmed` check which iterates through all a TX's inputs. I suppose I could move all that stuff out of `verify()` and into a new function `checkConflicts()` or something? Waddaya think?

tynes replied:
> The if does cut off a lot of unnecessary checks however

Good looks. I was trying to figure out how to move some code around so that `verify` no longer had a return value (like before this PR), but was worried that it would cause some performance issues around double checking things

pinheadmz replied:
So, you think it's ok as is?

pinheadmz replied:
so you think its ok as is?

pinheadmz commented:
Also want to call attention to this:

https://github.com/bcoin-org/bcoin/blob/7b47913f3c92cc661c9bcf3af6ab619db46e551e/lib/primitives/tx.js#L966-L969

This is NOT a protocol rule. TX version 2 is defined by [BIP 112](https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki) (OP_CSV) to add new functionality to the `nSequence` field -- but it does not "disable" RBF.

I think this line was added so that bcoin (rejecting RBF by default) would NOT reject OP_CSV transactions. Maybe once the wallet is RBF enabled and if we decide to RBF acceptance to `true` by default, we can take this out.

IMPranshu reviewed (commented):
Done with the review of code implementation of mempool part.

IMPranshu commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+        }
+      }
+
+      // Replacement TXs must not evict/replace more than 100 descendants
```
I opened a PR for this if it looks ok.
https://github.com/IMPranshu/bcoin-rbf/commit/07fc15d9ff691129d930cbb4e3bf9f19cf3afaf0

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -1659,13 +1741,39 @@ class Mempool extends EventEmitter {
+   * @returns {Promise} - Returns (@link MempoolEntry}[].
+   */
+
+  getConflicts(tx) {
+    const conflicts = [];
+    const hashes = new BufferSet();
+
+    for (const {prevout} of tx.inputs) {
+      const {hash, index} = prevout;
+      const conflict = this.getSpent(hash, index);
+
+      if (conflict && !hashes.has(hash)) {
+        hashes.add(hash);
+        conflicts.push(conflict);
+      }
+    }
+
+    return conflicts;
+  }
```
why not just do this?
```suggestion
function getConflicts(tx) {
  const conflicts = new Set();
  
  for (const { prevout: { hash, index } } of tx.inputs) {
    const conflict = this.getSpent(hash, index);
    
    if (conflict) {
      conflicts.add(conflict);
    }
  }

  return Array.from(conflicts);
}
```

pinheadmz replied:
great patch, thanks

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+        if (totalEvictions > 100) {
+          throw new VerifyError(tx,
+          'nonstandard',
+          'too many potential replacements',
```
We should add logs here like core
```C++
return strprintf("rejecting replacement %s; too many potential replacements (%d > %d)\n",
                 txid.ToString(),
                 nConflictingCount,
                 MAX_REPLACEMENT_CANDIDATES);
```

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+        }
+
+        if (entry.getDeltaRate() <= conflict.getDeltaRate()) {
+          throw new VerifyError(tx,
+          'insufficientfee',
+          'insufficient fee',
+          0);
```
adding comments to make life easier
```C++
return strprintf("rejecting replacement %s; new feerate %s <= old feerate %s",
                 txid.ToString(),
                 replacement_feerate.ToString(),
                 original_feerate.ToString());
```

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+      // transactions that will be evicted if an ancestor is replaced.
+      // Thus the replacement "pays for the bandwidth" of all the conflicts.
+      if (entry.deltaFee <= conflictingFees) {
+        throw new VerifyError(tx,
+        'insufficientfee',
+        'insufficient fee',
+        0);
```
```C++
return strprintf("rejecting replacement %s, less fees than conflicting txs; %s < %s",
                 txid.ToString(), FormatMoney(replacement_fees), FormatMoney(original_fees));
```

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+      // still needs to cover it's own bandwidth.
+      const feeRemaining = entry.deltaFee - conflictingFees;
+      if (feeRemaining < minFee) {
+        throw new VerifyError(tx,
+          'insufficientfee',
+          'insufficient fee',
+          0);
```
```C++
return strprintf("rejecting replacement %s, not enough additional fees to relay; %s < %s",
                 txid.ToString(),
                 FormatMoney(additional_fees),
                 FormatMoney(relay_fee.GetFee(replacement_vsize)));
```

theanmolsharma commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+      for (const {prevout} of tx.inputs) {
+        if (!this.isSpent(prevout.hash, prevout.index) &&
+            this.map.has(prevout.hash)) {
+          throw new VerifyError(tx,
+            'nonstandard',
+            'replacement-adds-unconfirmed',
+            0);
```
```C++
return strprintf("replacement %s adds unconfirmed input, idx %d",
                 tx.GetHash().ToString(), j);
```

theanmolsharma reviewed (commented):
Finished a first-pass code review. Nice work.
Left some comments about adding descriptive logs in addition to throwing errors.
Apart from the comments, I have 2 suggestions:

1. Should we preserve the order of checks like Core? I don't have a strong opinion on this. As long as we catch all the cases, any order should be fine but it's definitely something worth looking into.
2. I think it would be a good idea to move all RBF checks to a different function like Core's `MemPoolAccept::ReplacementChecks`

nodech commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+        totalEvictions += this.countDescendants(conflict) + 1;
+
+        // Replacement TXs must not evict/replace more than 100 descendants
+        if (totalEvictions > 100) {
```
Maybe move 100 into policy and use const from there for the 100.

nodech commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+      // Replacement TX must also pay for the total fees of all descendant
+      // transactions that will be evicted if an ancestor is replaced.
+      // Thus the replacement "pays for the bandwidth" of all the conflicts.
+      if (entry.deltaFee <= conflictingFees) {
```
This check is not strictly necessary as the next check covers this as well. If deltaFee <= conflictingFees the `feeRemaining` will be negative.

nodech commented on lib/mempool/mempool.js:
```diff
@@ -953,6 +967,72 @@ class Mempool extends EventEmitter {
+          0);
+      }
+
+      // Replacement TXs can not consume any unconfirmed outputs that were not
```
Not request to change, but I used it to better see what was happening, so basically asking if I understood what's up here:

```js
      // Replacement TXs can not consume any unconfirmed outputs that were not
      // already included in the original transactions. They can only add
      // confirmed UTXO, saving the trouble of checking new mempool ancestors.
      // Rule #2 checks
      for (const {prevout} of tx.inputs) {
        // We don't have the transaction in the mempool and it is not an
        // orphan, it means we are spending from the chain, no issues with
        // Rules #2.
        if (!this.map.has(prevout.hash))
          continue;

        // If the prevout is in the mempool and it was already being spent
        // by other transactions (which we are in conflict with) - we
        // are not violating rule #2.
        if (this.isSpent(prevout.hash, prevout.index))
          continue;

        // TX is spending new unconfirmed coin, which violates #2.
        throw new VerifyError(tx,
          'nonstandard',
          'replacement-adds-unconfirmed',
          0);
      }
```

nodech reviewed (commented):
It will be nice if you link the mempool-replacements policy from the Core and also document based on Rules (e.g. this does Rule 3 checks etc) for easier cross-checking.

pinheadmz commented:
@IMPranshu @nodech @theanmolsharma thanks so much for your reviews. I addressed all the comments and rebased on master in latest push. Also added a new commit that covers an edge case I hadn't thought of before: if we replace a tx that was spending an unconfirmed mempool output, the mempool would not "find" that coin and not only consider the replacement an orphan but also not even save it as an orphan.

nodech commented on lib/mempool/mempool.js:
```diff
@@ -513,8 +513,17 @@ class Mempool extends EventEmitter {
+    // If the TX that spent this coin signals
+    // replaceability, then even though it is "spent"
+    // we still "have" it.
+    if (spender) {
```
I think `hasCoin` becomes a not-ideal name for this if it depends on many things. Also this will ruin compatibility with the `getCoin` - even though it is not used for anything important.

One solution to this (and others I will mention) could be to pass new flag "rbf" to make it old behavior the default one and only use rbf when necessary and make it clearer. Alternatively, change the name of the method ?

nodech commented on lib/mempool/mempool.js:
```diff
@@ -1659,13 +1781,35 @@ class Mempool extends EventEmitter {
   isDoubleSpend(tx) {
     for (const {prevout} of tx.inputs) {
       const {hash, index} = prevout;
-      if (this.isSpent(hash, index))
```
Similar to above, I think it's good to have pure double spend vs rbf double spent methods separate or behind a flag. This is not critical in this specific method, but just a note I wanted to mention.

nodech reviewed (commented):
After another review and with the recent fixes to the rbf getting weirldy orphaned (but not really orphaned) - I noticed that `hasCoin` in the CoinView construction enables transactions that are not valid and currently can't be caught.

First let's start with the test case description:
```
  TX_A -> 2 outs
    TX_B -> spends TX_A.out[0]
      TX_C -> spends TX_B.out[0]
  TX_D -> spends TX_A.out[0]
    -> spends TX_B.out[0]
```

This is the example of the transaction (TX_D) that becomes valid and avoids CoinView/isDoubleSpend checks but is invalid. Test case will be provided below.

The situation occurs because TX_D does not really contradict any of the RBF rules itself:
  - It does not spend form NEW unconfirmed things, TX_B.out[0] and TX_A.out[0] are both spent already so this rule is good.
  Fees and other stuff is properly managed.

But because the coins in the CoinView are not retroactively removed, when TX_A, TX_B, TX_C are prepared for removal, TX_D verification passes. After that stuff breaks in following code when we are trying to evict entries (but that's beside the point here)

So ideally, when we are processing RBFs, we want to retroactively remove the coins and transactions from the CoinView (and ideally we should have similar abstraction as bitcoin where they have processing state with all necessary info that can be reverted easily, CoinView is the only thing that does that, mempool txs/spents etc are not easily modifiable on the fly like coinview is) - and do proper verification checks that Coins still exist for the new Transaction after "prepared to get removed" transactions have been subtracted from the coinview.

Test case
---
```js
    // Recursive bad:
    //   A -> 2 outs
    //     B -> spends A.out[0]
    //       C -> spends B.out[0]
    //   D -> spends A.out[0]
    //     -> spends B.out[0]
    it('should not accept for recursive spends', async () => {
      mempool.options.replaceByFee = true;

      const addr1 = chaincoins.createReceive().getAddress();
      const coin0 = chaincoins.getCoins()[0];

      // Generate parent TX
      const mtx0 = new MTX();
      mtx0.addCoin(coin0);
      mtx0.addOutput(addr1, coin0.value - 200);
      chaincoins.sign(mtx0);
      assert(mtx0.verify());
      const tx0 = mtx0.toTX();
      await mempool.addTX(tx0);

      const mtx1 = new MTX();
      const coin1 = Coin.fromTX(tx0, 0, -1);
      mtx1.addCoin(coin1);
      mtx1.inputs[0].sequence = 0xfffffffd;
      mtx1.addOutput(addr1, coin1.value - 200);
      chaincoins.sign(mtx1);
      assert(mtx1.verify());
      const tx1 = mtx1.toTX();
      await mempool.addTX(tx1);

      const mtx2 = new MTX();
      const coin2 = Coin.fromTX(tx1, 0, -1);
      mtx2.addCoin(coin2);
      mtx2.inputs[0].sequence = 0xfffffffd;
      mtx2.addOutput(addr1, coin2.value - 200);
      chaincoins.sign(mtx2);
      assert(mtx2.verify());
      const tx2 = mtx2.toTX();
      await mempool.addTX(tx2);

      // Send attempted replacement
      const mtx3 = new MTX();
      mtx3.addCoin(coin1);
      mtx3.addCoin(coin2);
      mtx3.addOutput(addr1, coin2.value + coin1.value - 1000);
      chaincoins.sign(mtx3);
      assert(mtx3.verify());
      const tx3 = mtx3.toTX();

      let err;

      try {
        await mempool.addTX(tx3);
      } catch (e) {
        err = e;
      }

      assert(err);
      console.log(err);
      // await assert.rejects(async () => {
      // }, {
      //   type: 'VerifyError',
      //   reason: 'bad-txns-inputs-spent'
      // });
    });
```

This currently hits assertion error in "unrelated" place:
```
AssertionError [ERR_ASSERTION]: Assertion failed.
    at Mempool.untrackEntry (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:1963:5)
    at Mempool.removeEntry (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:1219:10)
    at Mempool.evictEntry (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:1238:10)
    at Mempool.insertTX (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:867:14)
    at async Mempool._addTX (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:715:17)
    at async Mempool.addTX (/home/nd/dev/bcoin-org/bcoin/lib/mempool/mempool.js:694:14)
    at async Context.<anonymous> (/home/nd/dev/bcoin-org/bcoin/test/mempool-test.js:1621:9) {
  type: 'AssertionError',
  code: 'ERR_ASSERTION',
  generatedMessage: true,
  actual: false,
  expected: true,
  operator: '=='
}```

pinheadmz commented:
Awesome review @nodech and great catch! There was an ugly bug in Rule 2. I also addressed  your two other comments by restoring the existing implementations of `hasCoin()` and `isDoubleSpend()`. Details are in the commit messages, copied below.

## mempool: fix RBF Rule 2
    
    The current implementation improperly assumed that any unconfirmed
    inputs spent by the replacement TX were also spent by its direct
    conflict. It therefore did not account for the case where the
    replacement was spending an unconfirmed CHILD of its direct
    replacement. This would cause it to replace its own inputs which
    is an invalid mempool state and led to unexpected errors.

## mempool: restore original isDoubleSpend() implementation
    
    We only need to call the "quick and dirty test" if the user has
    set mempool RBF option to false. Do the Rule #1 check in verifyRBF()
    instead where it belongs. This leaves one loose end in maybeOrphan()
    that must be caught so we don't treat double-spends like orphans.

## mempool: allow replacement of tx spending unconfirmed utxo

This commit was refactored so the "is spender replaceable" logic is now in `getCoinView()` instead of
`hasCoin()`

Tests were updated with ASCII art as well ;-)

nodech commented on lib/mempool/mempool.js:
```diff
@@ -1058,6 +1059,36 @@ class Mempool extends EventEmitter {
+        // must also be spent by the conflict.
+        for (const {prevout: conflictPrevout} of conflict.tx.inputs) {
+          if (conflictPrevout.hash.equals(prevout.hash)
+              && conflictPrevout.input === prevout.input) {
```
Should this be `.index`? `.input` does not exist on `Outpoint`.

pinheadmz replied:
🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 

Fixed this in a new commit and added a test that fails without the patch
